lib/compat/wordpress-6.0/class-wp-webfonts.php

```diff
@@ -51,12 +80,288 @@ public function init() {
 			$this->stylesheet_handle = 'webfonts';
 			$hook                    = 'wp_enqueue_scripts';
 		}
+
+		add_action( 'init', array( $this, 'register_current_template_filter' ) );
+		add_action( 'init', array( $this, 'register_globally_used_webfonts' ) );
+
+		add_action( 'switch_theme', array( $this, 'invalidate_all_used_webfonts_cache' ) );
+		add_action( 'save_post_wp_template', array( $this, 'invalidate_used_webfonts_cache' ) );
+		add_action( 'save_post_wp_template_part', array( $this, 'invalidate_used_webfonts_cache' ) );
+
+		add_filter( 'the_content', array( $this, 'register_webfonts_used_in_content' ) );
+
+		add_filter( 'rest_request_after_callbacks', array( $this, 'maybe_invalidate_globally_used_webfonts_cache' ), 10, 3 );
+
 		add_action( $hook, array( $this, 'generate_and_enqueue_styles' ) );
 
 		// Enqueue webfonts in the block editor.
 		add_action( 'admin_init', array( $this, 'generate_and_enqueue_editor_styles' ) );
 	}
 
+	/**
+	 * Invalidates cache on global styles update.
+	 *
+	 * @param WP_REST_Response $response The response class.
+	 * @param WP_REST_Server   $handler The request handler.
+	 * @param WP_REST_Request  $request The request class.
+	 *
+	 * @return WP_REST_Response
+	 */
+	public function maybe_invalidate_globally_used_webfonts_cache( $response, $handler, $request ) {
+		if ( 'GET' === $request->get_method() || ! is_a( $handler['callback'][0], 'WP_REST_Global_Styles_Controller' ) ) {
+			// We only want to intercept requests that update global styles!
+			return $response;
+		}
+
+		$this->invalidate_globally_used_webfonts_cache();
+		update_option( self::$global_webfont_cache_option, $this->get_globally_used_webfonts() );
+
+		return $response;
+	}
+
+	/**
+	 * Register globally used webfonts.
+	 *
+	 * @return void
+	 */
+	public function register_globally_used_webfonts() {
+		$globally_used_webfonts = get_option( self::$global_webfont_cache_option );
+
+		if ( $globally_used_webfonts ) {
+			self::$used_webfonts = array_merge( self::$used_webfonts, $globally_used_webfonts );
+			return;
+		}
+
+		$globally_used_webfonts = $this->get_globally_used_webfonts();
+		update_option( self::$global_webfont_cache_option, $globally_used_webfonts );
+
+		self::$used_webfonts = array_merge( self::$used_webfonts, $globally_used_webfonts );
+	}
+
+	/**
+	 * Hook into every possible template so we can get the full template object on page load.
+	 *
+	 * @return void
+	 */
+	public function register_current_template_filter() {
+		$templates = get_block_templates( array(), 'wp_template' );
+
+		foreach ( $templates as $template ) {
+			add_filter(
+				$template->slug . '_template',
+				function() use ( $template ) {
+					$this->register_used_webfonts_by_template( $template );
+				}
+			);
+		}
+	}
```

Comment:
@creativecoder Do you happen to know if there's a single hook that returns the current template instead of what we're doing here?

Reply:
The `$template_html` global is available in `wp_head`, and contains all the HTML of the page if that helps (in block themes)

Reply:
Does it contain the slug too?

Reply:
> Does it contain the slug too?

No, `$template_html` is set in the `template-canvas.php` file when building the page, and only contains the HTML for the page:
https://github.com/WordPress/gutenberg/blob/f2e5e95eac1394c673f264f89c3ab69bf76c3f2e/lib/compat/wordpress-5.9/template-canvas.php#L12

Reply:
Got it, so it won't really work for our use case since we need the slug too.